server/user_db.go

```diff
@@ -0,0 +1,235 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package server
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/golang/protobuf/ptypes"
+
+	pb "github.com/googleapis/gapic-showcase/server/genproto"
+	"google.golang.org/genproto/protobuf/field_mask"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+func NewUserDb() UserDb {
+	return &userDb{
+		uid:   &uniqID{},
+		token: NewTokenGenerator(),
+		mu:    sync.Mutex{},
+		keys:  map[string]dbIndex{},
+		users: []userEntry{},
+	}
+}
+
+type UserDb interface {
+	Create(u *pb.User) (*pb.User, error)
+	Get(name string) (*pb.User, error)
+	Update(u *pb.User, f *field_mask.FieldMask) (*pb.User, error)
+	Delete(name string) error
+	List(page_size int32, token string) (*pb.ListUsersResponse, error)
+}
+
+type userEntry struct {
+	user    *pb.User
+	deleted bool
+}
+
+type dbIndex int
+
+type userDb struct {
+	uid   *uniqID
+	token TokenGenerator
+	mu    sync.Mutex
+	keys  map[string]dbIndex
+	users []userEntry
+}
+
+func (db *userDb) Create(u *pb.User) (*pb.User, error) {
+	db.mu.Lock()
+	defer db.mu.Unlock()
+
+	err := db.validate(u)
+	if err != nil {
+		return nil, err
+	}
+
+	// Validate Unique Fields.
+	hasSameName := func(x *pb.User) bool {
+		return u.GetDisplayName() == x.GetDisplayName()
+	}
+	if db.anyUser(hasSameName) {
+		return nil, status.Errorf(
+			codes.AlreadyExists,
+			"A user with display_name %s already exists.",
+			u.GetDisplayName())
+	}
+
+	hasSameEmail := func(x *pb.User) bool { return u.GetEmail() == x.GetEmail() }
+	if db.anyUser(hasSameEmail) {
+		return nil, status.Errorf(
+			codes.AlreadyExists,
+			"A user with the email %s already exists.",
+			u.GetEmail())
+	}
+
+	// Assign info.
+	id := db.uid.id()
+	name := fmt.Sprintf("users/%d", id)
+	now := ptypes.TimestampNow()
+
+	u.Name = name
+	u.CreateTime = now
+	u.UpdateTime = now
+
+	// Insert.
+	index := dbIndex(len(db.users))
+	db.users = append(db.users, userEntry{user: u, deleted: false})
+	db.keys[name] = index
+
+	return u, nil
+}
+
+func (db *userDb) Get(s string) (*pb.User, error) {
+	if v, ok := db.keys[s]; ok {
+		entry := db.users[int(v)]
+		if !entry.deleted {
+			return entry.user, nil
+		}
+	}
+
+	return nil, status.Errorf(codes.NotFound, "A user with name %s not found.", s)
+}
+
+func (db *userDb) Update(u *pb.User, f *field_mask.FieldMask) (*pb.User, error) {
+	if f != nil && len(f.GetPaths()) > 0 {
+		return nil, status.Error(
+			codes.Unimplemented,
+			"Field masks are currently not supported.")
+	}
+
+	db.mu.Lock()
+	defer db.mu.Unlock()
+
+	err := db.validate(u)
+	if err != nil {
+		return nil, err
+	}
+
+	if i, ok := db.keys[u.GetName()]; ok {
+		entry := db.users[int(i)]
+		if !entry.deleted {
+			// Validate Unique Fields.
+			hasSameName := func(x *pb.User) bool {
```

Comment:
Maybe factor this into `uniqNameEmail`?

Reply:
Done, but we do lose some granularity in the error message since we won't be able to specify which field had a collision.

Reply:
I don't have a strong opinion on this but something like this would also work

```
func (db *DB) VerifyNoDupe(name, email string) error {
  for _, u := range db.users {
    if u.Name == name {return errors.E("duplicate name")}
    if u.Email == email {return errors.E("duplicate email")}
  }
  return nil
}
```
then we don't need to deal with function pointers.

Reply:
Yeah thats better! My javascript is showing 😁.